### PR TITLE
hpal run

### PR DESCRIPTION
**Pull request by devinivy**
`hpal run` allows a user to run custom commands defined by their server's hapi plugins.

```
  hpal run [--list] <cmd> [<cmd-options>]
    e.g. hpal run plugin-name:command-name
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -93,3 +97,83 @@ When `<config-item>` is also specified, the first list item within the matched `
...
+
+if (!module.parent) {
+    exports.deployment(true);
+}
+```
+
+##### Creating your own commands
+
+Any hapi plugin can create commands that are runnable with `hpal run`!  Commands are exposed to hpal using hapi's [`server.expose()`](https://github.com/hapijs/hapi/blob/master/API.md#server.expose()).  Inside your plugin `my-plugin` simply call `server.expose('commands', commands)`, where `commands` is an object,
+ - whose keys are command names.  The name `default` is reserved for the command `hpal run my-plugin`.  Camel-cased command names are converted to kebab-case, so if the key is `someCommand` then it is run using `hpal run my-plugin:some-command`.
```
Comment (wswoodruff) [inline README.md]: Do you want the language to show the casing like saying `camelCased` and `kebob-cased`?

Comment (wswoodruff) [reply README.md]: Oh reading the next sentence you show an example, nvm!

File: lib/commands/run.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                    const commands = Object.keys(plugin.commands || {}).map((cmdName) => {
+
+                        const name = pluginName.replace(/^hpal-/, '') + ((cmdName === 'default') ? '' : internals.kebabize(`:${cmdName}`));
+                        const description = internals.normalizeCommand(plugin.commands[cmdName]).description;
+
+                        return { name, description };
+                    });
+
+                    return collect.concat(commands);
+                }, []);
```
Comment (wswoodruff) [inline lib/commands/run.js]: Cool, nize 👌

File: lib/index.js [hunk 3]
```diff
@@ -17,22 +17,25 @@ exports.start = (options) => {
...
-    const ctx = { options, colors };
+    const ctx = { options, colors, output, DisplayError };
 
     output(''); // Just a newline to make some room
 
-    if (args instanceof Error) {
-        throw new DisplayError(`${internals.usage(ctx)}\n\n` + colors.red(args.message));
-    }
+    const commandAndPart = (args instanceof Error) ? options.argv[2] : args._[2];
+    const extraArgs = (commandAndPart === 'run') ? options.argv : args._;
```
Comment (wswoodruff) [inline lib/index.js]: Did you get the `_` as prop convention from anywhere? Just curious

File: lib/index.js [hunk 4]
```diff
@@ -80,19 +83,32 @@ exports.start = (options) => {
...
                 .then(() => {
 
                     output(colors.green(`New pal project created in ${Path.relative(cwd, dir)}`));
                 });
         }
+        case 'run': {
+
+            const cwd = options.cwd;
+            const list = args.list;
+            const cmd = extraArgs[3];
```
Comment (wswoodruff) [inline lib/index.js]: Do we want to allow any flags with these commands?

Comment (wswoodruff) [reply lib/index.js]: Oh the flags would come _after_ the `run plugin:cmdName` gotcha

File: lib/index.js (hunk 3 above)
Comment (devinivy) [reply lib/index.js]: It's part of the bossy output: https://github.com/hapijs/bossy/blob/master/test/index.js#L80.  It consists of all the leftover argvs after parsing-out options in the bossy definition.

File: lib/index.js (hunk 4 above)
Comment (devinivy) [reply lib/index.js]: Yup!

**Comment by devinivy** [comment]
I think we should stop the server gracefully if a `DisplayError` is thrown.

**Pull request merged**